kernel: Consolidate thread state checking functions

TaiJuWu opened the pull request:
This series of two patches performs minor cleanup and refactoring within the scheduler code.

The first patch improves code consistency and readability by:

1. Consolidating thread state functions into a shared header file.

2. Replacing a direct bitwise check with a new, more descriptive helper function.

The second patch simplifies logic by:

1. Replacing a negative condition (!z_is_thread_prevented_from_running()) with a more readable positive check (z_is_thread_ready()).

2. Removing a redundant CONFIG_SMP conditional check.

3. Add the missing #endif directive.

TaiJuWu committed:
kernel: Consolidate thread state checking functions

This patch moves `is_aborting()` and `is_halting()`
from `kernel/sched.c` to `kernel/include/kthread.h`
and renames them to `z_is_thread_aborting()` and `z_is_thread_halting()`,
for consistency with other internal kernel APIs.

It replaces the previous inline function definitions in `sched.c`
with calls to the new header functions. Additionally, direct bitwise
checks like `(thread->base.thread_state & _THREAD_DEAD) != 0U`
are updated to use the new `z_is_thread_dead()` helper function.
This enhances code readability and maintainability.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

peter-mitsis commented:
I think that you are right in that those routines should be co-located with the others of similar structure in k_thread.h. However, I think it is also worthwhile to move one step further and name them as `z_is_thread_aborting` and `z_is_thread_halting` to improve naming consistency.

TaiJuWu committed:
kernel: cleanup thread state checks and nunecessary CONFIG check

The commit replaces negative thread state checks with a new,
 more descriptivepositive check.
The expression `!z_is_thread_prevented_from_running()`
is updated to `z_is_thread_ready()` where appropriate, making
the code's intent clearer.

 Removes a redundant `IS_ENABLED(CONFIG_SMP)`, they are included #ifdef.

Finally, this patch add the missing `#endif` directive.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

peter-mitsis reviewed (commented):
1. The moving and renaming of both is_halting() and is_aborting() looks good.
2. Switching to z_is_thread_ready() looks good.
3. Removal of IS_ENABLED(CONFIG_SMP) looks good.
4. Removal of setting old_thread->switch_handle to NULL at the start of z_get_next_switch_handle() seems to be OK. @andyross - Can you confirm/reject this? I've thrown what I could at it, but I worry I might have missed something.
5. The 2nd commit message indicates that that some `#endif` directives were misplaced; I can only see comments added to them. @TaiJuWu - Is this a stale message? Or did the patch miss something?

TaiJuWu commented:
> 1. The moving and renaming of both is_halting() and is_aborting() looks good.
> 2. Switching to z_is_thread_ready() looks good.
> 3. Removal of IS_ENABLED(CONFIG_SMP) looks good.


> 4. Removal of setting old_thread->switch_handle to NULL at the start of z_get_next_switch_handle() seems to be OK. @andyross - Can you confirm/reject this? I've thrown what I could at it, but I worry I might have missed something.

After thinking again, I'd like to revert this change since this PR should be clean up, I would not touch logic in this PR.


> 5. The 2nd commit message indicates that that some `#endif` directives were misplaced; I can only see comments added to them. @TaiJuWu - Is this a stale message? Or did the patch miss something?

Yes, this this stale message, thanks for your point out.

peter-mitsis reviewed (approved):
+1 For the change.

> There are many documents describing the active thread's handle must be NULL and only be themself within context switch so I thinks we can change it to assert instead of assigning.

FYI, with respect to the active thread's handle being NULL--that is my understanding too. I expect that dropping the NULL assignment _should_ be OK. Though it has passed every scenario I have thrown at it, I worry that I might be missing one (as I have been bitten enough times in the past by missed corner cases). If you want to change it to an assert in a separate commit/PR, I would be ok with that.

TaiJuWu commented:
> +1 For the change.
> 
> > There are many documents describing the active thread's handle must be NULL and only be themself within context switch so I thinks we can change it to assert instead of assigning.
> 
> FYI, with respect to the active thread's handle being NULL--that is my understanding too. I expect that dropping the NULL assignment _should_ be OK. Though it has passed every scenario I have thrown at it, I worry that I might be missing one (as I have been bitten enough times in the past by missed corner cases). If you want to change it to an assert in a separate commit/PR, I would be ok with that.

Look like we missed something, after do that change and run 
`west build $1 -b qemu_riscv64/qemu_virt_riscv64/smp -t run tests/kernel/threads/thread_apis \
 -T kernel.threads.apis.pinonly`
 
 We can get fail test and at that moment the thread_handle is dummy thread.
 I really not sure the reason right now.

TaiJuWu commented on kernel/include/kthread.h:
```diff
@@ -98,6 +98,25 @@ static inline bool z_is_thread_pending(struct k_thread *thread)
 	return (thread->base.thread_state & _THREAD_PENDING) != 0U;
 }
 
+static inline bool z_is_thread_dead(struct k_thread *thread)
+{
+	return (thread->base.thread_state & _THREAD_DEAD) != 0U;
+}
+
+/* Return true if the thread is aborting, else false */
+static inline bool z_is_thread_aborting(struct k_thread *thread)
+{
+	return (thread->base.thread_state & _THREAD_ABORTING) != 0U;
+}
+
+/* Return true if the thread is aborting or suspending, else false */
+static inline bool z_is_thread_halting(struct k_thread *thread)
+{
+	return (thread->base.thread_state &
+		(_THREAD_ABORTING | _THREAD_SUSPENDING)) != 0U;
+}
```
There are many similar functions and I didn't see any test about them.
If these tests exist, could you help me to point out where are they and I'd happily to add tests.

kartben replied:
> There are many similar functions and I didn't see any test about them. If these tests exist, could you help me to point out where are they and I'd happily to add tests.

The rest of the file was pretty much 100% covered before this change, hence why I was mentioning it :) https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/kernel%2Finclude%2Fkthread.h

I am not a kernel expert so maybe @peter-mitsis @andyross can help give some hints - I would be really surprised if there isn't an existing test case that you can just extend. Thank you for the prompt follow up!
